File: Completed/JavaScript/451.js
/**
 * @param {string} s
 * @return {string}
 */
var frequencySort = function(s) {
  // counting
  
  const map = {}
  for (let char of s) {
      map[char] = (map[char] || 0) + 1
  }
  
  return Object.keys(map).sort((a,b) => map[b] - map[a]).reduce((str, key) => {
      for(let i = 0; i < map[key]; i++) {
          str += key
      }
      
      return str
  }, '')
};

File: Completed/Go/451.go
import (
    "strings"
    "sort"
)

func frequencySort(s string) string {
    arr := []string{}
    freq := map[string]int{}
    // get freqs O(N)
    for _, r := range s {
        str := string(r)
        freq[str]++
        
        arr = append(arr, str)
    }
    
    // heapify O(N log N)
    sort.Slice(arr, func(i, j int) bool {
        a, b := arr[i], arr[j]
        if freq[a] == freq[b] {
            return a > b
        }
        return freq[a] > freq[b]
    })
    
    // O(N)
    return strings.Join(arr, "")
}
